File: src/runtime/random.c
#include "random.h"

#include <time.h>

#define MT19937_N 624
#define MT19937_M 397
#define MT19937_MATRIX_A 0x9908B0DFu
#define MT19937_UPPER_MASK 0x80000000u
#define MT19937_LOWER_MASK 0x7FFFFFFFu

static void random_twist(random_t* random) {
    for (uint32_t i = 0; i < MT19937_N; i++) {
        uint32_t y = (random->mt[i] & MT19937_UPPER_MASK) |
                     (random->mt[(i + 1) % MT19937_N] & MT19937_LOWER_MASK);
        uint32_t next = random->mt[(i + MT19937_M) % MT19937_N] ^ (y >> 1);
        if (y & 1u) {
            next ^= MT19937_MATRIX_A;
        }
        random->mt[i] = next;
    }
    random->index = 0;
}

void random_seed(random_t* random, uint32_t seed) {
    if (!random) {
        return;
    }

    random->mt[0] = seed;
    for (uint32_t i = 1; i < MT19937_N; i++) {
        uint32_t prev = random->mt[i - 1];
        random->mt[i] = 1812433253u * (prev ^ (prev >> 30)) + i;
    }
    random->index = MT19937_N;
}

void random_init(random_t* random) {
    if (!random) {
        return;
    }
    uint32_t seed = (uint32_t)time(NULL);
    random_seed(random, seed);
}

uint32_t random_get(random_t* random) {
    if (!random) {
        return 0;
    }

    if (random->index >= MT19937_N) {
        random_twist(random);
    }

    uint32_t y = random->mt[random->index++];

    y ^= (y >> 11);
    y ^= (y << 7) & 0x9D2C5680u;
    y ^= (y << 15) & 0xEFC60000u;
    y ^= (y >> 18);

    return y;
}

uint32_t random_range(random_t* random, uint32_t max) {
    if (!random || max == 0) {
        return 0;
    }
    return random_get(random) % max;
}

File: fri3d-kernel/src/random.rs
const MT19937_N: usize = 624;
const MT19937_M: usize = 397;
const MT19937_MATRIX_A: u32 = 0x9908B0DF;
const MT19937_UPPER_MASK: u32 = 0x80000000;
const MT19937_LOWER_MASK: u32 = 0x7FFFFFFF;

pub struct Random {
    mt: [u32; MT19937_N],
    index: usize,
}

impl Random {
    pub fn new(seed: u32) -> Self {
        let mut rng = Self {
            mt: [0; MT19937_N],
            index: MT19937_N,
        };
        rng.seed(seed);
        rng
    }

    pub fn seed(&mut self, seed: u32) {
        self.mt[0] = seed;
        for i in 1..MT19937_N {
            let prev = self.mt[i - 1];
            self.mt[i] = 1812433253u32.wrapping_mul(prev ^ (prev >> 30)).wrapping_add(i as u32);
        }
        self.index = MT19937_N;
    }

    pub fn get(&mut self) -> u32 {
        if self.index >= MT19937_N {
            self.twist();
        }

        let mut y = self.mt[self.index];
        self.index += 1;

        y ^= y >> 11;
        y ^= (y << 7) & 0x9D2C5680;
        y ^= (y << 15) & 0xEFC60000;
        y ^= y >> 18;

        y
    }

    pub fn range(&mut self, max: u32) -> u32 {
        if max == 0 {
            return 0;
        }
        self.get() % max
    }

    fn twist(&mut self) {
        for i in 0..MT19937_N {
            let y = (self.mt[i] & MT19937_UPPER_MASK)
                | (self.mt[(i + 1) % MT19937_N] & MT19937_LOWER_MASK);
            let mut next = self.mt[(i + MT19937_M) % MT19937_N] ^ (y >> 1);
            if y & 1 != 0 {
                next ^= MT19937_MATRIX_A;
            }
            self.mt[i] = next;
        }
        self.index = 0;
    }
}
